### PR TITLE
Add tuple instances for `Simulate`, `Drivable`, `Backpressure` and `Test`

### PR DESCRIPTION
**Pull request by lmbollen**
To allow simulating tuples of circuits

### CODE REVIEW

**Review by martijnbastiaan — APPROVED** [review]

**Review by rowanG077 — APPROVED** [review]

File: clash-protocols/src/Protocols/Internal.hs [hunk 1]
```diff
@@ -88,25 +95,13 @@ infixr 1 |>
...
   boolsToBwd _ bs = (boolsToBwd (Proxy @a) bs, boolsToBwd (Proxy @b) bs)
 
-instance (Backpressure a, Backpressure b, Backpressure c) => Backpressure (a, b, c) where
-  boolsToBwd _ bs =
-    ( boolsToBwd (Proxy @a) bs
-    , boolsToBwd (Proxy @b) bs
-    , boolsToBwd (Proxy @c) bs
-    )
-
+backPressureTupleInstances 3 maxTupleSize
```
Comment (DigitalBrains1) [inline clash-protocols/src/Protocols/Internal.hs:104]:
```suggestion
backPressureTupleInstances 3 maxTupleSize

```

Please add a blank line

File: clash-protocols/src/Protocols/Internal.hs [hunk 2]
```diff
@@ -422,10 +263,7 @@ instance (Drivable a, Drivable b) => Drivable (a, b) where
       , sampleC @b conf (Circuit $ \_ -> ((), fwd2))
       )
 
--- TODO TemplateHaskell?
--- instance SimulateType (a, b, c)
--- instance SimulateType (a, b, c, d)
-
+drivableTupleInstances 3 maxTupleSize
```
Comment (DigitalBrains1) [inline clash-protocols/src/Protocols/Internal.hs:266]:
```suggestion
drivableTupleInstances 3 maxTupleSize

```

Please add a blank line

**Review by DigitalBrains1 — CHANGES_REQUESTED** [review]
You forgot a backtick in the commit summary:
> Move `BackPressure`, `Simulate`, `Drivable` and `Test

File: clash-protocols/src/Protocols/Internal.hs [hunk 3]
```diff
@@ -33,8 +34,14 @@ import qualified Clash.Explicit.Prelude as CE
 import Clash.Prelude (type (*), type (+))
 import qualified Clash.Prelude as C
 
+import Protocols.Internal.TH (
+  backPressureTupleInstances,
+  drivableTupleInstances,
+  simulateTupleInstances,
+ )
```
[the remark this answers is not in the document]
Comment (lmbollen) [reply clash-protocols/src/Protocols/Internal.hs:41]: This style is enforced by our `fourmulu` configuration

File: clash-protocols/src/Protocols/Internal/TH.hs [hunk 4]
```diff
@@ -31,3 +36,167 @@ idleCircuitTupleInstance n =
...
+   where
+    mkStallPat =
+      tupP
+        [ varP (mkName $ "stalls" <> show i)
+        , varP (mkName $ if i == n then "_" else "rem" <> show i)
+        ]
+
+  -- stallC Declaration: Generate stalling circuits
+  stallCircuits = zipWith mkStallCircuit [1 .. n] circTys
+  mkStallCircuit i ty =
```
[the remark this answers is not in the document]
Comment (lmbollen) [reply clash-protocols/src/Protocols/Internal/TH.hs:92]: Unfortunately I couldnt figure out how :(

File: clash-protocols/src/Protocols/Internal.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (DigitalBrains1) [reply clash-protocols/src/Protocols/Internal.hs:41]: Huh... weird.

File: clash-protocols/src/Protocols/Internal.hs (hunk 2 above)
Comment (DigitalBrains1) [reply clash-protocols/src/Protocols/Internal.hs:267]: Since this PR was merged without this suggestion, I added it to #124.

**Pull request merged**